### PR TITLE
fix pipeline deprecations for redis-rb 4.6+

### PR DESCRIPTION
**Pull request by skipkayhil**
Ref #1794 

This is by no means a complete fix of all the warnings (yet), but I figured I could knock out some of the obvious ones and start a discussion about the less obvious ones

### CODE REVIEW

**Comment by morenocarullo** [comment]
This is great, what can I do to help out here?

**Comment by pb-dennis-marchand** [comment]
Would love to see this get merged!

File: lib/resque/data_store.rb [hunk 1]
```diff
@@ -99,10 +99,11 @@ class QueueAccess
       def initialize(redis)
         @redis = redis
       end
+
       def push_to_queue(queue,encoded_item)
-        @redis.pipelined do
-          watch_queue(queue)
-          @redis.rpush redis_key_for_queue(queue), encoded_item
+        @redis.pipelined do |pipeline|
+          pipeline.sadd(:queues, queue.to_s) # watch_queue
```
[the remark this answers is not in the document]
Comment (nevans) [reply lib/resque/data_store.rb:105]: IMO, less invasive than this and better than creating a new QueueAccess instance: any pipeline-able methods should receive an optional `redis:` keyword argument.  E.g.

```ruby
      def push_to_queue(queue, encoded_item, redis: @redis)
        redis.pipelined do |pipeline|
          watch_queue(queue, redis: pipeline)
          pipeline.rpush redis_key_for_queue(queue), encoded_item
        end
      end

      def watch_queue(queue, redis: @redis)
        redis.sadd(:queues, queue.to_s)
      end
```

That's more or less the approach I took with my own application code, and should work anywhere the code is compatible with being pipelined... and _presumably_ all code currently being run inside a pipeline should be compatible with being pipelined.  :)

**Comment by morenocarullo** [comment]
@jrochkind -- you can silence logs by adding `Redis.silence_deprecations = true` in an appropriate initializer.

**Comment by nevans** [comment]
FWIW, I finally got around to tackling this gem for my own upgrades.  I just posted #1806, which removes *all* deprecation warnings from the tests but is failing the 2.3 and 2.4 builds due to the way I used kwargs.  If there's interest, I can update it to work with 2.3 and 2.4.

Thanks @skipkayhil for getting this started!

**Comment by iloveitaly** [comment]
@nevans right now we are compatible with 2.3 + 2.4, so if we can keep that compatibility, let's try.

**Comment by matthewhively** [comment]
I suppose this can be closed now that #1806 is merged

**Comment by iloveitaly** [comment]
I think so, although it seems like this PR is a decent bit more complex than the one that was merged. I'm curious to compare and see if anything was missed that is covered in this PR—if anyone can do that, I'd appreciate it!

**Comment by nevans** [comment]
FWIW, I think this can be closed.  I think you'll find that #1806 updates _more_ methods in the code, but with a smaller change to each method.  So long as you pass the pipeline object down to every method that could possibly need it (as an optional arg that defaults to the connection object) I don't think it's necessary to inline any methods or check any arity.  See my comment here: https://github.com/resque/resque/pull/1796#discussion_r890624193  I'd probably write it a little bit differently if I were refactoring or writing the classes from scratch, but I was intentionally aiming for the "simplest thing that will *safely* work".  I've been using #1806 in production since before it was merged, and `deprecation_toolkit` hasn't noticed anything.

And again, thanks to @skipkayhil for pushing this PR first!

**Comment by iloveitaly** [comment]
Thanks @nevans! @PatrickTulskie seems to agree here too. Closing this out!

**Pull request closed without merging**